packages/aila/src/protocol/jsonPatchProtocol.ts

```diff
@@ -282,108 +411,305 @@ export const JsonPatchDocumentJsonSchema = zodToJsonSchema(
+    }
+    return { parsed, isPartial: false };
+  } catch (e) {
+    // If parsing fails, assume it's partial
+    return { parsed: null, isPartial: true };
+  }
 }
 
-interface PotentialMessage {
-  type?: string;
-  content?: string;
+export function tryParsePart(
+  obj: object,
+): MessagePartDocument | UnknownDocument {
+  const { type } = obj as { type: string };
+  // Assert the message part type is allowed
+  if (!MessagePartDocumentSchemaByType[type as MessagePartType]) {
+    console.error("Invalid message part type", type);
+    return {
+      type: "unknown",
+      value: JSON.stringify,
+      error: "Invalid message part type",
+    };
+  }
+  // Parse the object with the correct schema
+  const parsed =
+    MessagePartDocumentSchemaByType[type as MessagePartType].safeParse(obj);
+  if (parsed.success) {
+    return parsed.data;
+  } else {
+    return { type: "unknown", value: JSON.stringify(obj), error: parsed.error };
+  }
 }
-interface PartialMessage {
-  type?: string;
-  path?: string;
-  isPartial: boolean;
-  value?:
-    | {
-        path?: string;
-        op?: string;
-        value?: unknown;
-      }
-    | string;
+
+export function tryParsePatch(obj: object): PatchDocument | UnknownDocument {
+  const parsed = PatchDocumentSchema.safeParse(obj);
+  if (parsed.success) {
+    const patchDocument: PatchDocument = parsed.data;
+    return patchDocument;
+  } else {
+    console.log("Unable to parse patch", parsed, parsed.error);
+    return { type: "unknown", value: JSON.stringify(obj), error: parsed.error };
+  }
 }
-export function parseMessagePart(part: string): PartialMessage | undefined {
-  const trimmed = part.trim();
-  if (!trimmed.startsWith("{")) {
-    return { type: "text", value: trimmed, isPartial: false };
+
+function tryParseText(obj: object): TextDocument | UnknownDocument {
+  const parsed = TextDocumentSchema.safeParse(obj);
+  if (parsed.success) {
+    return parsed.data;
+  } else {
+    return { type: "unknown", value: JSON.stringify(obj), error: parsed.error };
   }
-  const isPartial = !trimmed.endsWith("}");
-  const untruncated = untruncateJson(trimmed);
-  const parsed: PotentialMessage = parseJsonSafely(untruncated);
-  if (typeof parsed !== "object") {
-    return undefined;
+}
+
+// Each Message that is sent back from the server contains the following
+// (separated by the record-separator character and a newline):
+// * An llmMessage matching the LLMMessageSchema, and containing multiple messageParts
+// * A moderation messagePart
+// * An ID messagePart
+// * A state messagePart
+// (Potentially more)
+// So we split the message into rows and parse each one.
+// The message is streaming in, so we evaluate each row to see if it is
+// valid JSON. If it is not, then we assume that the row is streaming
+// and set an isPartial boolean on the resulting object.
+// This helps us not to re-process past messageParts after they have
+// fully streamed in.
```

Comment:
It's useful to have all the jsonPatchProtocol work in the same file for the PR as it makes a useful diff. I think we should follow up with a PR to split out the schema definition from the utility functions. Comments like this are really useful and at the moment get a bit lost in the file